secret_store/src/key_server_cluster/cluster_connections_net.rs

```diff
@@ -0,0 +1,543 @@
+	/// Mutable connection data.
+	container: Arc<RwLock<NetConnectionsContainer>>,
+}
+
+/// Network connections container. This is the only mutable data of NetConnectionsManager.
+/// The set of nodes is mutated by the connection trigger and the connections set is also
+/// mutated by spawned futures.
+pub struct NetConnectionsContainer {
+	/// Is this node isolated from cluster?
+	pub is_isolated: bool,
+	/// Current key servers set.
+	pub nodes: BTreeMap<NodeId, SocketAddr>,
+	/// Active connections to key servers.
+	pub connections: BTreeMap<NodeId, Arc<NetConnection>>,
+}
+
+/// Network connection to single key server node.
+pub struct NetConnection {
+	executor: Executor,
+	/// Id of the peer node.
+	node_id: NodeId,
+	/// Address of the peer node.
+	node_address: SocketAddr,
+	/// Is this inbound (true) or outbound (false) connection?
+	is_inbound: bool,
+	/// Key pair that is used to encrypt connection' messages.
+	key: KeyPair,
+	/// Last message time.
+	last_message_time: RwLock<Instant>,
+	/// Underlying TCP stream.
+	stream: SharedTcpStream,
+}
+
+impl NetConnectionsManager {
+	/// Create new network connections manager.
+	pub fn new(
+		executor: Executor,
+		message_processor: Arc<MessageProcessor>,
+		trigger: Box<ConnectionTrigger>,
+		container: Arc<RwLock<NetConnectionsContainer>>,
+		config: &ClusterConfiguration,
+		net_config: NetConnectionsManagerConfig,
+	) -> Result<Self, Error> {
+		let listen_address = make_socket_address(
+			&net_config.listen_address.0,
+			net_config.listen_address.1)?;
+
+		Ok(NetConnectionsManager {
+			listen_address,
+			data: Arc::new(NetConnectionsData {
+				allow_connecting_to_higher_nodes: net_config.allow_connecting_to_higher_nodes,
+				executor: executor,
+				message_processor,
+				self_key_pair: config.self_key_pair.clone(),
+				trigger: Mutex::new(trigger),
+				container,
+			}),
+		})
+	}
+
+	/// Start listening for connections and schedule connections maintenance.
+	pub fn start(&self) -> Result<(), Error> {
+		net_listen(&self.listen_address, self.data.clone())?;
+		net_schedule_maintain(self.data.clone());
+		Ok(())
+	}
+}
+
+impl ConnectionManager for NetConnectionsManager {
+	fn provider(&self) -> Arc<ConnectionProvider> {
+		self.data.container.clone()
+	}
+
+	fn connect(&self) {
+		net_connect_disconnected(self.data.clone());
+	}
+}
+
+impl ConnectionProvider for RwLock<NetConnectionsContainer> {
+	fn connected_nodes(&self) -> Result<BTreeSet<NodeId>, Error> {
+		let connections = self.read();
+		if connections.is_isolated {
+			return Err(Error::NodeDisconnected);
+		}
+
+		Ok(connections.connections.keys().cloned().collect())
+	}
+
+	fn disconnected_nodes(&self) -> BTreeSet<NodeId> {
+		let connections = self.read();
+		connections.nodes.keys()
+			.filter(|node_id| !connections.connections.contains_key(node_id))
+			.cloned()
+			.collect()
+	}
+
+	fn connection(&self, node: &NodeId) -> Option<Arc<Connection>> {
+		match self.read().connections.get(node).cloned() {
```

Comment:
Do you really need to pattern match in this?

Isn't `self.read().connections.get(node).cloned()` sufficient?